### PR TITLE
2283 Add Breaches tab and breach list view in Monitoring page

### LINKED ISSUE #2283: Add `Breaches` tab and breach list view in `Monitoring` page
## Is your feature request related to a problem? Please describe 👀
Cold chain needs a way to view breaches in the UI.

## Describe the solution you'd like 🎁
- The monitoring page has 3 tabs ( only implementing the breach tab in this issue ) 
- Hook to query `temperature_logs` and ( hook to update temperature log comment - comment not yet implemented tho...)
- Columns: 
  - Status. Icon for unacknowledged breaches: Group 125
  - Sensor name
  - Storage location
  - CCE
  - Breach start (datetime)
  - Breach end (datetime)
  - Duration (minutes)
  - Ongoing
  - Hot or cold. Cold icon: 🦆 icon _snow flake_ Hot icon: 🦆 icon _sun light_
  - Highest/Lowest Temperature
  - Type
  - Comment ( will need backend/mSupply issue first to implement )

UI mockup: 
Screenshot 2023-10-12 at 12 01 05

### Estimate
6 hours

### Moneyworks Jobcode 🧰
OMS:COLC

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #2283 

# 👩🏻‍💻 What does this PR do? 
Adds a way for the user to view a list of breaches.
Screenshot 2023-10-17 at 18 17 30

# 🧪 How has/should this change been tested? 
- [ ] Go to monitoring
- [ ] Click on the breaches tab and see a list view of breaches
- [ ] Can also try to change tabs to see if the views still load

## 💌 Any notes for the reviewer?
Have left out the comment column for now...

## 📃 Documentation
Overall cold chain documentation

### CODE REVIEW

File: client/packages/common/src/intl/locales/en/common.json [hunk 1]
```diff
@@ -9,33 +9,32 @@
   "button.add-item": "Add Item",
   "button.allocate": "Allocate",
   "button.cancel": "Cancel",
-  "button.save-and-confirm-status": "Confirm {{status}}",
   "button.close": "Close",
+  "button.create": "Create",
```
Comment (roxy-dao) [inline client/packages/common/src/intl/locales/en/common.json:13]: Sorry... this was just putting everything in alphabetical order... I know @mark-prins likes the file this way 😉 ...

File: client/packages/common/src/utils/formatters/formatters.ts [hunk 2]
```diff
@@ -30,11 +30,16 @@ export const Formatter = {
   ): string => Papa.unparse(data, config),
   csvDateTimeString: (dateString?: string | null | undefined): string => {
     const date = dateString ? new Date(dateString) : null;
-    return date && isValid(date) ? format(date, "yyyy-MM-dd' 'HH:mm:ss") : '';
+    return date && isValid(date) ? format(date, "dd/MM/yyyy' 'HH:mm:ss") : '';
```
Comment (roxy-dao) [inline client/packages/common/src/utils/formatters/formatters.ts:33]: Requested change from Richard and also consistent with other parts of UI..

File: server/graphql/types/src/types/temperature_breach.rs [hunk 3]
```diff
@@ -144,6 +144,15 @@ impl TemperatureBreachNode {
             .await?
             .map(LocationNode::from_domain))
     }
+
+    pub async fn temperature_log(&self, ctx: &Context<'_>) -> Result<Option<TemperatureLogNode>> {
```
Comment (roxy-dao) [inline server/graphql/types/src/types/temperature_breach.rs:148]: Needed this for the temperature...

File: server/repository/src/db_diesel/temperature_breach.rs [hunk 4]
```diff
@@ -85,7 +85,7 @@ impl<'a> TemperatureBreachRepository<'a> {
                 }
             }
         } else {
-            query = query.order(temperature_breach_dsl::start_datetime.asc())
+            query = query.order(temperature_breach_dsl::start_datetime.desc())
```
Comment (roxy-dao) [inline server/repository/src/db_diesel/temperature_breach.rs:88]: Sorry... I kept getting confused with asc and desc for some reason 🤦🏼

**Comment by roxy-dao** [comment]
Not sure why the breach wasn't linked to a log... is there a bug in the cold chain app?

File: server/graphql/types/src/types/temperature_breach.rs (hunk 3 above)
Comment (jmbrunskill) [reply server/graphql/types/src/types/temperature_breach.rs:148]: It seems like the design is to record the breach_id against all temperature log records related (so there might actually be a whole lot of associated temperature log records)

I'm guessing the design is intending to show either the max temperature or the min temperature (depending on the breach type?

image

File: client/packages/coldchain/src/Monitoring/api/TemperatureBreach/operations.graphql [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  sensor {
+    ...Sensor
+  }
+
+  location {
+    ...LocationRow
+  }
+
+  temperatureLog {
```
Comment (jmbrunskill) [inline client/packages/coldchain/src/Monitoring/api/TemperatureBreach/operations.graphql:18]: I would say it's probably better to return `min_temperature` and `max_temperature`, give more flexibility in how it's queried and allows you to display the right data.

The temperatureLogs should really be an array of related temperature records, which could be displayed in a graph or something in future.

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        sortable: false,
+      },
+      {
+        key: 'temperature',
+        label: 'label.temperature',
+        accessor: ({ rowData }) => {
+          return `${rowData.temperatureLog?.temperature}${t(
+            'label.temperature-unit'
+          )}`;
+        },
```
Comment (mark-prins) [inline client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx:108]: I had to do this:

```suggestion
        accessor: ({ rowData }) => {
          return `${rowData.sensor?.latestTemperatureLog?.nodes[0]
            ?.temperature}${t('label.temperature-unit')}`;
        },
```

maybe I have bad data?

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+  const columns = useColumns<TemperatureBreachFragment>(
+    [
+      {
+        key: 'acknowledgedIcon',
+        Cell: ({ rowData }) => {
+          return !rowData?.acknowledged ? (
+            <CircleAlertIcon
+              fill={theme.palette.error.main}
+              sx={{ color: '#fff' }}
```
Comment (mark-prins) [inline client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx:41]: this is possibly not helpful. just wondered if we could remove a hard coded something.
```suggestion
              sx={{ color: 'background.white' }}
```

File: client/packages/coldchain/src/common/BreachTypeCell.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  return (
+    <>
+      {breachType && (
+        <Typography
+          sx={{
+            color:
+              temperature === 'HOT'
+                ? theme.palette.warning.main
+                : theme.palette.secondary.dark,
```
Comment (mark-prins) [inline client/packages/coldchain/src/common/BreachTypeCell.tsx:33]: this really is just a suggestion.. an alternative to using `useTheme`
```suggestion
            color: temperature === 'HOT' ? 'warning.main' : 'secondary.dark',
```

and also

```
          {temperature === 'HOT' ? (
            <SunIcon sx={{ stroke: 'warning.main' }} />
          ) : (
            <SnowflakeIcon sx={{ stroke: 'secondary.dark' }} />
          )}
```

File: client/packages/common/src/utils/formatters/formatters.ts (hunk 2 above)
Comment (mark-prins) [reply client/packages/common/src/utils/formatters/formatters.ts:33]: hm.

File: server/graphql/types/src/types/temperature_breach.rs (hunk 3 above)
Comment (mark-prins) [reply server/graphql/types/src/types/temperature_breach.rs:148]: yeah, my preference would be to return just the temperature.

**Review by mark-prins — APPROVED** [review]
👍 LGTM!

fantastic to see the data come in! looks great - my only concern is the sort and that my dataset isn't working proper.

File: client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx (hunk 6 above)
Comment (roxy-dao) [reply client/packages/coldchain/src/Monitoring/ListView/TemperatureBreach/TemperatureBreachList.tsx]: There were some foreign key constraints during sync.. have fixed this hopefully!

File: client/packages/coldchain/src/common/BreachTypeCell.tsx (hunk 8 above)
Comment (mark-prins) [reply client/packages/coldchain/src/common/BreachTypeCell.tsx:33]: Sorry, it wasn't clear - there's a two part suggestion here and the first part isn't resolved - and if you make that change, you can remove `useTheme` 😄

**Pull request merged**